### ISSUE TITLE
optimizing tfb.MaskedAutoregressiveFlow: what to pass as var_list to optimizer.minimize() in TF 2?

### ISSUE THREAD

**Issue by skeydan** [tensorflow/probability#520]
Hi,

 I had a unit test for (Inverse) MaskedAutoregressiveFlow that looked like this (see below) and worked well with TF 1. For TF 2, `optimizer.minimize` needs to be passed a `var_list` in addition to the `loss`.

As e.g. `iaf.trainable_variables` is an empty list, I am wondering what to pass here?

Many thanks in advance!

```
iaf = tfd.TransformedDistribution(
      distribution=tfd.Normal(loc=0., scale=1.),
      bijector=tfb.Invert(tfb.MaskedAutoregressiveFlow(
          shift_and_log_scale_fn=tfb.masked_autoregressive_default_template(
              hidden_layers=[512, 512]))),
      event_shape=[5])

target_dist = tfd.Normal(loc = 2.2, scale = 0.23)
y = target_dist.sample(1000)
y = tf.reshape(y, shape = [200, 5])
loss = lambda _: - tf.reduce_mean(iaf.log_prob(y))
    
optimizer = tf.optimizers.Adam(1e-4)
optimizer.minimize(loss)

```

**Comment by MokkeMeguru** [comment]
I'm not a developer, but I faced on a similar problem, so I show you my solution.
Please see. #545 

I can advice 
1. wrap (around) MaskedAutoregressiveflow  by DistributionLambda 
2. [wrapped bijector].vars = bijector.trainable_variables 

But this main poblem is tfp may uses tf-1's style yet… (see. #355) I think.

I hope you can solve your problem.

**Comment by skeydan** [comment]
Thanks! I was asking because of the unit tests in the TFP for R wrapper ().

I think I should just disable the old-style tests and add new ones using `tfb.AutoregressiveNetwork` as recommended here:

https://groups.google.com/a/tensorflow.org/d/msg/tfprobability/jg6iXMD1Nw4/Sm_4BAKnAgAJ

BTW looking forward to your GLOW example! :-)

**Comment by MokkeMeguru** [comment]
@skeydan 
(Here is the glow's repo (https://github.com/MokkeMeguru/glow-realnvp-tutorial)/ issue (#554))

**Comment by skeydan** [comment]
Thanks! Do you still plan on sending a PR to the TFP examples?

At some later time, I might be interested to port it to R :-))

**Comment by MokkeMeguru** [comment]
I don't know the manner of their PR... (https://github.com/tensorflow/probability/issues/554#issuecomment-532016853)
But I'll try it once this week.
